File: v2/golang/client/query_ast_live_errors_test.go
//go:build astlive

package client

// query_ast_live_errors_test.go — every query the grammar forbids, and the
// question that actually matters: does THIS SDK refuse it before a byte leaves
// the process, or does it send it and hand back the server's 400?
//
// Domain: the rejection contract. Two different contracts hide behind the same
// word "error", and the three SDKs are free to disagree about which one a given
// mistake gets.
//
// Junior Tip [why "who rejected it" is recorded, not just "it errored",
// 2026-09-14]: a client-side refusal costs no round trip and can carry a
// message the server never wrote; a server-side refusal costs a request and
// carries the server's exact bytes. A caller that retries on Retryable(), or
// that matches on the message text, behaves differently in the two worlds. The
// `wantClientSide` column below is therefore an assertion about the CONTRACT,
// and `wireWasEmpty` is its proof: nothing reached the tap.

import (
	"errors"
	"fmt"
	"strings"
	"testing"
)

// astErrorCase is one forbidden query and the contract it is expected to meet.
type astErrorCase struct {
	name             string
	builderCall      string
	build            func(fixture *astFixture) *QueryRequest
	wantClientSide   bool
	wantErrSubstring string
	wantHTTPStatus   int       // meaningful only when the request reached the server
	wantKind         ErrorKind // meaningful only when the request reached the server
	note             string
}

// astErrorCases is the full forbidden-query matrix for the Go SDK.
var astErrorCases = []astErrorCase{
	{
		name:             "unknown filter column",
		builderCall:      `scoped.Where("bogus_col", QueryOp{Eq: 1})`,
		build:            func(fixture *astFixture) *QueryRequest { return scoped(fixture).Where("bogus_col", QueryOp{Eq: 1}) },
		wantClientSide:   true,
		wantErrSubstring: `query: field "bogus_col" is not allowed in filters`,
	},
	{
		name:             "filter column in the wrong case",
		builderCall:      `scoped.Where("TYPE", QueryOp{Eq: "fact"})`,
		build:            func(fixture *astFixture) *QueryRequest { return scoped(fixture).Where("TYPE", QueryOp{Eq: "fact"}) },
		wantClientSide:   true,
		wantErrSubstring: `query: field "TYPE" is not allowed in filters`,
	},
	{
		name:        "consolidate_id — a real column the grammar does not expose",
		builderCall: `scoped.Where("consolidate_id", QueryOp{Eq: 1})`,
		build: func(fixture *astFixture) *QueryRequest {
			return scoped(fixture).Where("consolidate_id", QueryOp{Eq: 1})
		},
		wantClientSide:   true,
		wantErrSubstring: `query: field "consolidate_id" is not allowed in filters`,
	},
	{
		name:        "semantic_search pseudo-key — the SERVER accepts and skips it, Go refuses it",
		builderCall: `scoped.Where("semantic_search", QueryOp{Eq: "anything"})`,
		build: func(fixture *astFixture) *QueryRequest {
			return scoped(fixture).Where("semantic_search", QueryOp{Eq: "x"})
		},
		wantClientSide:   true,
		wantErrSubstring: `query: field "semantic_search" is not allowed in filters`,
		note:             "DIVERGENCE: server returns 200 for this key (record_ast_query.go:169-172); Go cannot send it at all",
	},
	{
		name:             "operator object with no operator set",
		builderCall:      `scoped.Where("type", QueryOp{})`,
		build:            func(fixture *astFixture) *QueryRequest { return scoped(fixture).Where("type", QueryOp{}) },
		wantClientSide:   true,
		wantErrSubstring: `query: filter "type": no operator survived encoding`,
	},
	{
		name:        "$eq with an explicit nil — indistinguishable from unset in Go",
		builderCall: `scoped.Where("superseded_by", QueryOp{Eq: nil})`,
		build: func(fixture *astFixture) *QueryRequest {
			return scoped(fixture).Where("superseded_by", QueryOp{Eq: nil})
		},
		wantClientSide:   true,
		wantErrSubstring: `query: filter "superseded_by": no operator survived encoding`,
		note: "NOT A GAP (amended 2026-09-14): QueryOp.Eq is interface{} with omitempty so `$eq: null` is unreachable " +
			"from Go, and that is CORRECT. Live 2026-09-14: {\"superseded_by\":{\"$eq\":null}} answers 200 count=0 on a " +
			"tenant whose unfiltered page returns 1000 rows, every one of which satisfies superseded_by IS NULL — the " +
			"server compiles it to `col = NULL`, never true. Python/TS can send a dead predicate; Go cannot.",
	},
	{
		name:        "$in with an empty list",
		builderCall: `scoped.Where("type", QueryOp{In: []interface{}{}})`,
		build: func(fixture *astFixture) *QueryRequest {
			return scoped(fixture).Where("type", QueryOp{In: []interface{}{}})
		},
		wantClientSide:   true,
		wantErrSubstring: `query: filter "type": $in requires a non-empty list of values`,
	},
	{
		name:             "sort on a column outside the whitelist",
		builderCall:      `scoped.OrderBy("bogus", "asc")`,
		build:            func(fixture *astFixture) *QueryRequest { return scoped(fixture).OrderBy("bogus", "asc") },
		wantClientSide:   true,
		wantErrSubstring: `query: sort field "bogus" is not allowed`,
	},
	{
		name:             "sort with an unrecognised direction — the SERVER silently uses DESC",
		builderCall:      `scoped.OrderBy("id", "sideways")`,
		build:            func(fixture *astFixture) *QueryRequest { return scoped(fixture).OrderBy("id", "sideways") },
		wantClientSide:   true,
		wantErrSubstring: `query: sort order "sideways" is not allowed`,
		note:             "DIVERGENCE: server answers 200 and orders DESC (record_ast_query.go:240-242); Go refuses",
	},
	{
		name:             "limit above the server cap — the SERVER silently clamps to 1000",
		builderCall:      `scoped.Limit(1001)`,
		build:            func(fixture *astFixture) *QueryRequest { return scoped(fixture).Limit(1001) },
		wantClientSide:   true,
		wantErrSubstring: "query: limit must be between 1 and 1000, got 1001",
		note:             "DIVERGENCE: server clamps silently (astQueryMaxLimit); Go refuses, so the clamp is unreachable from Go",
	},
	{
		name:             "limit zero — the SERVER falls back to 50",
		builderCall:      `scoped.Limit(0)`,
		build:            func(fixture *astFixture) *QueryRequest { return scoped(fixture).Limit(0) },
		wantClientSide:   true,
		wantErrSubstring: "query: limit must be between 1 and 1000, got 0",
		note:             "DIVERGENCE: server applies the default 50; Go refuses",
	},
	{
		name:             "negative offset — the SERVER coerces to 0",
		builderCall:      `scoped.Offset(-1)`,
		build:            func(fixture *astFixture) *QueryRequest { return scoped(fixture).Offset(-1) },
		wantClientSide:   true,
		wantErrSubstring: "query: offset cannot be negative, got -1",
		note:             "DIVERGENCE: server coerces to 0 and answers 200; Go refuses",
	},
	{
		name:        "$eq with an OBJECT value — Go does NOT check scalarity, the server does",
		builderCall: `scoped.Where("type", QueryOp{Eq: map[string]interface{}{"a": 1}})`,
		build: func(fixture *astFixture) *QueryRequest {
			return scoped(fixture).Where("type", QueryOp{Eq: map[string]interface{}{"a": 1}})
		},
		wantClientSide:   false,
		wantErrSubstring: "operator $eq expects a single value",
		wantHTTPStatus:   400,
		wantKind:         KindInvalidRequest,
	},
	{
		name:        "$eq with an ARRAY value — also only caught server-side",
		builderCall: `scoped.Where("score", QueryOp{Eq: []interface{}{1, 2}})`,
		build: func(fixture *astFixture) *QueryRequest {
			return scoped(fixture).Where("score", QueryOp{Eq: []interface{}{1, 2}})
		},
		wantClientSide:   false,
		wantErrSubstring: "operator $eq expects a single value",
		wantHTTPStatus:   400,
		wantKind:         KindInvalidRequest,
	},
	{
		name:        "$in containing a nested array element",
		builderCall: `scoped.Where("type", QueryOp{In: []interface{}{[]string{"a"}}})`,
		build: func(fixture *astFixture) *QueryRequest {
			return scoped(fixture).Where("type", QueryOp{In: []interface{}{[]string{"a"}}})
		},
		wantClientSide:   true,
		wantErrSubstring: `query: filter "type": $in[0] is a []string — $in elements must be scalars`,
		note: "AMENDED 2026-09-14: this row expected the SERVER to reject it, and was already stale when written — the same commit added the client-side scalar guard that prevents the round trip. " +
			"The SERVER fact it recorded is still true: a nested array inside $in answers HTTP 400 \"expects a single value\". Go cannot observe that any more (Validate() runs on every path, " +
			"no raw-AST escape hatch); Python and TypeScript still prove it with a hand-written AST (their query_ast_live_null_contract suites).",
	},
	{
		name:        "request body above the 1 MiB REST cap",
		builderCall: `scoped.Where("summary", QueryOp{In: 20000 long strings})`,
		build: func(fixture *astFixture) *QueryRequest {
			padding := strings.Repeat("x", 64)
			values := make([]interface{}, 0, 20000)
			for element := 0; element < 20000; element++ {
				values = append(values, fmt.Sprintf("%s-%d", padding, element))
			}
			return scoped(fixture).Where("summary", QueryOp{In: values})
		},
		wantClientSide:   false,
		wantErrSubstring: "invalid json ast",
		wantHTTPStatus:   400,
		wantKind:         KindInvalidRequest,
		note:             "the oversize message is SHORTER than the parse-failure message — a client matching on the long text will not recognise it",
	},
}

// runASTErrorMatrix exercises the rejection contract and records every case.
func runASTErrorMatrix(testHandle *testing.T, harness *astLiveHarness, fixture *astFixture) {
	testHandle.Helper()
	for _, errorCase := range astErrorCases {
		currentCase := errorCase
		testHandle.Run(currentCase.name, func(subTest *testing.T) {
			outcome := runASTCase(subTest, harness, "error/"+currentCase.name,
				currentCase.builderCall, currentCase.build(fixture))

			if outcome.queryErr == nil {
				subTest.Fatalf("expected a rejection, got %d rows (status %d, wire=%s)",
					len(outcome.raw), outcome.status, outcome.wireBody)
			}
			errorText := outcome.queryErr.Error()
			if !strings.Contains(errorText, currentCase.wantErrSubstring) {
				subTest.Errorf("error text %q does not contain %q", errorText, currentCase.wantErrSubstring)
			}

			wireWasEmpty := outcome.wireBody == ""
			if currentCase.wantClientSide && !wireWasEmpty {
				subTest.Errorf("expected a CLIENT-SIDE refusal but %d bytes reached the server (status %d)",
					len(outcome.wireBody), outcome.status)
			}
			if !currentCase.wantClientSide {
				if wireWasEmpty {
					subTest.Fatalf("expected the request to REACH the server, but nothing left the process")
				}
				if outcome.status != currentCase.wantHTTPStatus {
					subTest.Errorf("server answered HTTP %d, expected %d", outcome.status, currentCase.wantHTTPStatus)
				}
				var apiError *APIError
				if !errors.As(outcome.queryErr, &apiError) {
					subTest.Fatalf("server rejection is not a *APIError (%T) — the caller cannot read Kind()/Retryable()",
						outcome.queryErr)
				}
				if apiError.Kind() != currentCase.wantKind {
					subTest.Errorf("Kind()=%q expected %q", apiError.Kind(), currentCase.wantKind)
				}
				if apiError.Retryable() {
					subTest.Errorf("Retryable()=true for a %d — a caller would loop on a permanent rejection", apiError.StatusCode)
				}
			}
			fmt.Printf("AST_ERROR name=%q client_side=%v wire_bytes=%d status=%d note=%q\n",
				currentCase.name, wireWasEmpty, len(outcome.wireBody), outcome.status, currentCase.note)
		})
	}
}

// assertClientSideErrorsAreTypedLikePythonAndTypeScript pins the parity Go
// reached with newValidationError: a refusal the BUILDER makes is the same
// *APIError the transport path returns, so one errors.As covers both origins.
//
// AMENDED 2026-09-14: this used to assert the OPPOSITE (bare fmt.Errorf, no
// status), with its own escape clause — "if someone later makes Validate()
// return a typed error, this test goes red and the parity note gets updated
// instead of rotting". That happened; the note is updated rather than the
// assertion loosened, because the divergence is GONE.
//
// Junior Tip [the type alone is not the contract]: callers branch on errors.As,
// then on Kind() and Retryable(). A typed error with the WRONG Kind is worse
// than an untyped one — it looks classifiable and classifies wrong. So all
// three are asserted, plus clientSide: a refusal that reached the server is not
// a refusal, it is a complaint filed after the fact.
func assertClientSideErrorsAreTypedLikePythonAndTypeScript(testHandle *testing.T, fixture *astFixture) {
	testHandle.Helper()
	request := scoped(fixture).Where("bogus_col", QueryOp{Eq: 1})
	validationErr := request.Validate()
	if validationErr == nil {
		testHandle.Fatal("Validate accepted a column outside the whitelist")
	}
	var apiError *APIError
	if !errors.As(validationErr, &apiError) {
		testHandle.Fatalf("client-side validation returned %T, not *APIError — a caller branching "+
			"on errors.As cannot classify it, the divergence this test closes", validationErr)
	}
	if apiError.Kind() != KindInvalidRequest {
		testHandle.Errorf("client-side rejection Kind() = %v, want %v", apiError.Kind(), KindInvalidRequest)
	}
	if apiError.Retryable() {
		testHandle.Error("a malformed query must never report Retryable() true")
	}
	if !apiError.clientSide {
		testHandle.Error("clientSide must be true — the 400 is a prediction of what the server " +
			"WOULD have answered, not a status it did answer")
	}
	fmt.Printf("AST_CONTRACT client_side_error_is_typed=true err_type=%T kind=%v retryable=%v\n",
		validationErr, apiError.Kind(), apiError.Retryable())
}

// assertNilRequestIsRefused covers the degenerate call.
func assertNilRequestIsRefused(testHandle *testing.T, harness *astLiveHarness) {
	testHandle.Helper()
	harness.resetSnapshot()
	records, queryErr := harness.memory.Query(testContext(), nil)
	if queryErr == nil {
		testHandle.Fatalf("Query(nil) returned %d records instead of an error", len(records))
	}
	if !strings.Contains(queryErr.Error(), "request is required") {
		testHandle.Errorf("Query(nil) error %q does not name the missing request", queryErr.Error())
	}
	wireBody, _, _ := harness.snapshot()
	if wireBody != "" {
		testHandle.Errorf("Query(nil) still put %d bytes on the wire", len(wireBody))
	}
}

File: v2/typescript/src/test/query_ast_live_errors.test.ts
/**
 * LIVE: the AST error contract, and the mistakes the server ACCEPTS in silence.
 *
 * Two different contracts live here and a caller must not confuse them:
 *
 *   1. REJECTED — HTTP 400, surfaced as `AnhurQueryError` with
 *      kind "invalid_request", retryable false, statusCode 400. The server's
 *      own sentence is embedded (escaped) in `.message`.
 *   2. ACCEPTED AND IGNORED — HTTP 200 with a result set that is not the one
 *      the caller thinks they asked for. These are the dangerous ones: a
 *      failure wearing a success status. Every case in the second block was
 *      confirmed against production; each is a place where a reasonable person
 *      would have predicted a 400 and does not get one.
 *
 *     ANHUR_AST_LIVE=1 ANHUR_API_KEY=... npm test
 */
import { after, before, describe, it } from "node:test";
import * as assert from "node:assert/strict";
import { AnhurQueryError } from "../types.js";
import {
  ascending, LIVE_ENABLED, LIVE_SKIP_REASON, seedAstFixture, serverReason,
  teardownAstFixture, type AstFixture,
} from "./astLiveFixture.js";

let fixture: AstFixture;

/** Assert the server rejects `ast` with the full typed-error contract. */
async function serverRejects(ast: Record<string, unknown>, reason: RegExp): Promise<void> {
  await assert.rejects(
    () => fixture.memory.query(ast as never),
    (rejection: AnhurQueryError) => {
      assert.equal(rejection instanceof AnhurQueryError, true, "must be AnhurQueryError");
      assert.equal(rejection.statusCode, 400);
      assert.equal(rejection.kind, "invalid_request");
      assert.equal(rejection.retryable, false, "a malformed query must never be retried");
      assert.match(serverReason(rejection), reason);
      return true;
    },
  );
}

describe("live AST — error contract", { skip: LIVE_ENABLED ? false : LIVE_SKIP_REASON }, () => {
  before(async () => { fixture = await seedAstFixture("errors"); });
  after(async () => { await teardownAstFixture(fixture); });

  describe("rejections the SERVER owns (the builder cannot pre-empt them)", () => {
    it("rejects a column outside the whitelist", async () => {
      await serverRejects({ filters: { bogus_col: { $eq: 1 } } }, /invalid filter field: "bogus_col"/);
    });

    it("rejects an operator it does not implement", async () => {
      await serverRejects({ filters: { type: { $ne: "fact" } } }, /unsupported operator "\$ne"/);
    });

    it("rejects a bare value where an operator object belongs", async () => {
      await serverRejects({ filters: { type: "fact" } }, /must be an object of operators/);
    });

    it("rejects an operator-less filter — the shape an `undefined` value produces", async () => {
      // `whereEquals("type", options.type)` with an absent option builds
      // `{"type":{}}`, because JSON.stringify drops undefined. The builder does
      // not catch it; this 400 is the only feedback the caller gets.
      await serverRejects({ filters: { type: {} } }, /has no operator/);
    });

    it("rejects a non-scalar value, naming the Go type it received", async () => {
      await serverRejects({ filters: { type: { $eq: { a: 1 } } } }, /got map\[string\]interface \{\}/);
      await serverRejects({ filters: { type: { $eq: ["fact"] } } }, /got \[\]interface \{\}/);
    });

    it("rejects an empty $in list — a boundary the builder lets through", async () => {
      await serverRejects({ filters: { type: { $in: [] } } }, /\$in requires a non-empty array of values/);
    });

    it("rejects a non-array $in value", async () => {
      await serverRejects({ filters: { type: { $in: "fact" } } }, /\$in requires a non-empty array of values/);
    });

    it("rejects a nested array inside $in", async () => {
      await serverRejects({ filters: { type: { $in: [["fact"]] } } }, /got \[\]interface \{\}/);
    });

    it("rejects top-level limit/offset instead of silently ignoring them", async () => {
      // This used to be an accept-and-ignore: a caller asking for 500 rows
      // quietly got 50. It is now a 400 precisely so the mistake is visible.
      await serverRejects({ limit: 5, filters: {} }, /must live inside the "pagination" object/);
      await serverRejects({ offset: 5, filters: {} }, /must live inside the "pagination" object/);
    });

    it("rejects a bad sort field, including the empty one a missing key produces", async () => {
      await serverRejects({ sort: [{ field: "bogus", order: "asc" }] }, /invalid sort field: "bogus"/);
      await serverRejects({ sort: [{ order: "asc" }] }, /invalid sort field: ""/);
    });

    it("rejects an oversized body with a DIFFERENT, shorter message", async () => {
      // Junior Tip [do not string-match the long message]: a body over 1 MiB
      // fails at MaxBytesReader and answers the bare 'invalid json ast',
      // WITHOUT the grammar hint the undecodable-body case carries. A client
      // matching the long text will not recognise this one.
      const oversize = {
        filters: {
          uuid: { $eq: fixture.mainSession },
          summary: { $in: Array.from({ length: 12000 }, (_, index) => "x".repeat(100) + index) },
        },
      };
      await assert.rejects(
        () => fixture.memory.query(oversize as never),
        (rejection: AnhurQueryError) => {
          assert.equal(rejection.statusCode, 400);
          assert.equal(serverReason(rejection), "invalid json ast");
          return true;
        },
      );
    });
  });

  describe("ACCEPTED AND IGNORED — HTTP 200 where a 400 would be kinder", () => {
    it("parses `select` and then ignores it, returning whole records", async () => {
      const result = await fixture.memory.query(
        fixture.scoped().select("id").orderBy("id", "asc").build(),
      );
      assert.deepEqual(result.records.map((record) => record.id), fixture.visibleIds);
      assert.ok(result.records[0].summary, "select:['id'] must NOT strip the other columns");
    });

    it("ignores an unknown top-level key entirely", async () => {
      const result = await fixture.memory.query({
        unknown_top: "x",
        filters: { uuid: { $eq: fixture.mainSession } },
        pagination: { limit: 1000 },
      } as never);
      assert.deepEqual(ascending(result.records.map((record) => record.id)), fixture.visibleIds);
    });

    it("silently sorts DESC on an unrecognised sort direction", async () => {
      // The direction whitelist is ASC/DESC/asc/desc; anything else falls back
      // to DESC without a word. TypeScript's union type is the only thing
      // stopping this, and a cast or a JavaScript caller walks straight past it.
      const result = await fixture.memory.query({
        filters: { uuid: { $eq: fixture.mainSession } },
        sort: [{ field: "id", order: "sideways" }],
        pagination: { limit: 1000 },
      } as never);
      assert.deepEqual(
        result.records.map((record) => record.id),
        [...fixture.visibleIds].reverse(),
      );
    });

    it("silently CLAMPS an oversized limit rather than rejecting it", async () => {
      const result = await fixture.memory.query({
        filters: { uuid: { $eq: fixture.mainSession } },
        pagination: { limit: 5000 },
      } as never);
      assert.equal(result.count, fixture.visibleIds.length);
    });

    it("silently replaces a non-positive limit with the default 50", async () => {
      const result = await fixture.memory.query({
        filters: { uuid: { $eq: fixture.mainSession } },
        pagination: { limit: 0 },
      } as never);
      assert.equal(result.count, fixture.visibleIds.length);
    });

    it("accepts and SKIPS the `semantic_search` pseudo-key, contributing nothing", async () => {
      // Not a column: the server logs it and moves on. A caller who believes
      // they added a semantic leg gets a purely structural query and no warning.
      const result = await fixture.memory.query({
        filters: { semantic_search: { query: "anything at all" }, uuid: { $eq: fixture.mainSession } },
        pagination: { limit: 1000 },
      } as never);
      assert.deepEqual(ascending(result.records.map((record) => record.id)), fixture.visibleIds);
    });

    it("TRAP: $eq null answers 200 with an empty page, on every column", async () => {
      // `null` is in the server's scalar set, so it compiles to `col = ?` bound
      // to NULL — never true in SQL. `superseded_by IS NULL` is pinned on EVERY
      // returnable row, so the honest answer is "all of them"; the caller gets
      // zero, with no error. Go cannot express this; TypeScript and Python can,
      // which makes the trap theirs alone.
      const result = await fixture.memory.query(
        fixture.scoped().whereEquals("superseded_by", null).build(),
      );
      assert.equal(result.count, 0);
      assert.deepEqual(result.records, []);
    });

    it("normalises the wire's `records: null` into an empty array", async () => {
      // The server returns a nil Go slice on zero hits, so the wire literally
      // carries "records": null — never []. Callers index this; the SDK must
      // absorb it or every empty result is a TypeError.
      const result = await fixture.memory.query(
        fixture.scoped().whereEquals("type", "no-such-type").build(),
      );
      assert.deepEqual(result.records, []);
      assert.equal(result.count, 0);
    });
  });
});
